clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -95,23 +92,11 @@
 
     private final Time time;
 
-    private final Map<Integer, Collection<ApiVersionsResponse.ApiVersion>> nodeApiVersions = new HashMap<>();
+    private final Map<Integer, NodeApiVersions> nodeApiVersions = new HashMap<>();
 
     private final Set<String> nodesNeedingApiVersionsFetch = new HashSet<>();
 
-    public NetworkClient(Selectable selector,
-                         Metadata metadata,
-                         String clientId,
-                         int maxInFlightRequestsPerConnection,
-                         long reconnectBackoffMs,
-                         int socketSendBuffer,
-                         int socketReceiveBuffer,
-                         int requestTimeoutMs,
-                         Time time,
-                         Collection<ApiVersionsResponse.ApiVersion> requiredApiVersions) {
-        this(NO_METADATA_UPDATER, metadata, selector, clientId, maxInFlightRequestsPerConnection,
-                reconnectBackoffMs, socketSendBuffer, socketReceiveBuffer, requestTimeoutMs, time, requiredApiVersions);
-    }
+    private final List<ClientResponse> abortedSends = new LinkedList<>();
```

Comment:
Well, the `ArrayList` array would expand to 10 elements the first time an aborted send is added. And then, it will probably never grow beyond that. LinkedList in Java is just not a very good collection in general. Each node has 3 references and there is a node instance per item. A funny tweet by Joshua Bloch:

"@jerrykuch @shipilev @AmbientLion Does anyone actually use LinkedList?  I wrote it, and I never use it."
https://twitter.com/joshbloch/status/583813919019573248